### ISSUE TITLE
Behavior of nested collections

### ISSUE THREAD

**Issue by andrewthad** [JustusAdam/mustache#6]
Here is an example of a template I have:

    {{#models}}
    data {{model_name.title}}' = {{model_name.title}}' {
      {{#fields}}
        {{model_name.title}} {{field_name.title}}
      {{/fields}}
      }   
    {{/models}}

This template is intended to generate part of a haskell data declaration. Here is the result:

    data Car' = Car' {
         Speed
         Legal
      }
    data Person' = Person' {
         Name
         Age
         Height
      }

I expected to see this:

    data Car' = Car' {
         Car Speed
         Car Legal
      }
    data Person' = Person' {
         Person Name
         Person Age
         Person Height
      }

The problem is that the fields made available by iterating over `models` seem to not be in scope in the `fields` subcollection. I haven't read the mustache spec, so I don't know if this behavior is intended.

**Comment by JustusAdam** [comment]
Hmm, curious ...

The problem with the spec is that it is essentially a collection of tests that run against the library and not actually formulated in words.

I however find your proposal very interesting and I'll have a try at implementing it, if it doesn't break the tests I'll merge it.

You should however also consider proposing this very issue/testcase to the [mustache spec repo][specs] as well. I am sure they'd be very much interested in adding test cases to their spec for such obviously currently undefined behaviour.
Should you choose to do so they'll most likely require you to provide a short description of the behaviour/test case, the template to be substituted, the expected result and a yaml or json data structure with the substituted values.
Basically the same structure they currently use (see the [repo itself][specs])
[specs]: https://github.com/mustache/spec

**Comment by andrewthad** [comment]
Thanks for looking at implementing it. I appreciate that a lot. I have opened up a ticket on the mustache spec repo: https://github.com/mustache/spec/issues/94. I have never heard of a spec being defined that way before, but I will say that something in me finds it intriguing.

**Comment by andrewthad** [comment]
I've created a test case that I'm going to try to get into the spec. If you'd like to try it out in the meantime, you can change `Spec.hs` to use my fork:

    langspecDir = "andrewthad-spec-786e4ac"
    -- langspecDir = "spec-1.1.3"
    specDir = "specs"
    releaseFile = "langspec.tar.gz"
    releaseURL = "https://codeload.github.com/andrewthad/spec/legacy.tar.gz/add_list_context_check"
    -- releaseURL = "https://codeload.github.com/mustache/spec/tar.gz/v1.1.3"

**Comment by JustusAdam** [comment]
All right. I ran the tests locally. They failed as expected. However I cannot yet find the source of the bug, I am suspecting an issue when changing context but I haven't been able to find it yet. Might take a while to track this down.

**Comment by JustusAdam** [comment]
You're very much welcome to help. I am suspecting the error to be somewhere in either `Text.Mustache.Render.substituteValue.substitute' (curried)` or `Text.Mustache.Render.search`

**Comment by andrewthad** [comment]
It's definitely in `search`. I'm in the process of rewriting it.

**Comment by andrewthad** [comment]
Got it.

**Comment by andrewthad** [comment]
In the PR, I also add a `Show` instance for `Context`, which let's you use `showTraceId`, which was really helpful for tracking down what went wrong.

**Comment by JustusAdam** [comment]
Yes that solved the problem. Thanks a lot. I hope the test makes it into the spec soon. For the time being I'll be running against your fork of the tests.
